### PR TITLE
Feature 11 create update items

### LINKED ISSUE #11: Support creating and updating entities

### PR DESCRIPTION
**Pull request by platonovsimeon**
Resolves #11

### CODE REVIEW

File: src/main/kotlin/fi/metatavu/odata/mock/data/DataProvider.kt [hunk 1]
```diff
@@ -43,6 +44,20 @@ class DataProvider {
         return result
     }
 
+    /**
+     * Reads entity by entry id
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/odata/mock/data/DataProvider.kt:48]: It seems to return entity, not entry

File: src/main/kotlin/fi/metatavu/odata/mock/processor/ODataProcessor.kt [hunk 2]
```diff
@@ -446,9 +462,75 @@ class ODataProcessor(private val dataProvider: DataProvider) : EntityCollectionP
...
+            val serializedContent = createSerializedEntity(
+                entry = Entry(id = entries.first(), name = entityName, data = entryData),
+                entitySet = edmEntitySet,
+                responseFormat = responseFormat,
+                uriInfo = uriInfo
+            ).content
+            response.content = serializedContent
+            response.statusCode = HttpStatusCode.OK.statusCode
+            response.setHeader(HttpHeader.CONTENT_TYPE, responseFormat.toContentTypeString())
+        }
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/odata/mock/processor/ODataProcessor.kt:495]:
```suggestion
        val entryData = String(request.body.readAllBytes())
        val edmEntitySet = getEdmEntitySet(uriInfo.asUriInfoResource())
        val idPropertyName = edmEntitySet.entityType.keyPropertyRefs.first().name
        val idPropertyValue = jacksonObjectMapper().readTree(entryData).get(idPropertyName)
        val entityName = edmEntitySet.entityType.name
        val entryIds = mutableListOf<UUID>()

        DataContainer.getEntries(entityName).forEach {
            val mockEntryId = it.id!!
            val foundEntity = DataProvider().readByEntryId(edmEntitySet, mockEntryId)!!
            val entryId = "${foundEntity.getProperty(idPropertyName).value}"
            val queryId = "$idPropertyValue".trim('"')

            if (entryId == queryId) {
                DataContainer.removeEntry(mockEntryId)
                entryIds.add(mockEntryId)
            }
        }

        if (entryIds.size == 0) {
            throw ODataApplicationException(
                "No entry found for this key", HttpStatusCode.NOT_FOUND
                    .statusCode, Locale.ENGLISH
            )
        } else {
            val serializedContent = createSerializedEntity(
                entry = Entry(id = entryIds.first(), name = entityName, data = entryData),
                entitySet = edmEntitySet,
                responseFormat = responseFormat,
                uriInfo = uriInfo
            ).content
            response.content = serializedContent
            response.statusCode = HttpStatusCode.OK.statusCode
            response.setHeader(HttpHeader.CONTENT_TYPE, responseFormat.toContentTypeString())
        }
```

File: src/main/kotlin/fi/metatavu/odata/mock/processor/ODataProcessor.kt [hunk 3]
```diff
@@ -193,10 +200,19 @@ class ODataProcessor(private val dataProvider: DataProvider) : EntityCollectionP
...
+        val entryName = entitySet.name
+        val entryData = String(request.body.readAllBytes())
+        val serializedContent = createSerializedEntity(
+            entry = Entry(name = entryName, data = entryData),
+            entitySet = entitySet,
+            responseFormat = responseFormat,
+            uriInfo = uriInfo
+        ).content
+
+        response.content = serializedContent
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/odata/mock/processor/ODataProcessor.kt:213]:
```suggestion
        DataContainer.addEntry(Entry(name = entryName, data = entryData))
        response.content = entryData.byteInputStream()
```

File: src/main/kotlin/fi/metatavu/odata/mock/processor/ODataProcessor.kt [hunk 4]
```diff
@@ -446,9 +462,75 @@ class ODataProcessor(private val dataProvider: DataProvider) : EntityCollectionP
...
+                    .statusCode, Locale.ENGLISH
+            )
+        } else {
+            val serializedContent = createSerializedEntity(
+                entry = Entry(id = entries.first(), name = entityName, data = entryData),
+                entitySet = edmEntitySet,
+                responseFormat = responseFormat,
+                uriInfo = uriInfo
+            ).content
+            response.content = serializedContent
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/odata/mock/processor/ODataProcessor.kt:492]:
```suggestion
            DataContainer.addEntry(Entry(id = entryIds.first(), name = entityName, data = entryData))
            response.content = entryData.byteInputStream()
```

File: src/main/kotlin/fi/metatavu/odata/mock/processor/ODataProcessor.kt [hunk 5]
```diff
@@ -446,9 +462,75 @@ class ODataProcessor(private val dataProvider: DataProvider) : EntityCollectionP
...
+        DataContainer.getEntries(entityName).forEach {
+            val foundEntity = DataProvider().readByEntryId(edmEntitySet, it.id!!)
+            if (foundEntity!!.getProperty(idPropertyName).value.toString() == idPropertyValue.toString()) {
+                val entryId = it.id
+                DataContainer.removeEntry(entryId)
+                entries.add(entryId)
+            }
+        }
+
+        if (entries.size == 0) {
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/odata/mock/processor/ODataProcessor.kt:480]:
```suggestion
        if (entryIds.size == 0) {
```

File: src/main/kotlin/fi/metatavu/odata/mock/processor/ODataProcessor.kt [hunk 6]
```diff
@@ -446,9 +462,75 @@ class ODataProcessor(private val dataProvider: DataProvider) : EntityCollectionP
...
+        val entityName = edmEntitySet.entityType.name
+        val entries = mutableListOf<UUID>()
+        DataContainer.getEntries(entityName).forEach {
+            val foundEntity = DataProvider().readByEntryId(edmEntitySet, it.id!!)
+            if (foundEntity!!.getProperty(idPropertyName).value.toString() == idPropertyValue.toString()) {
+                val entryId = it.id
+                DataContainer.removeEntry(entryId)
+                entries.add(entryId)
+            }
+        }
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/odata/mock/processor/ODataProcessor.kt:478]:
```suggestion
        val entryIds = mutableListOf<UUID>()

        DataContainer.getEntries(entityName).forEach {
            val mockEntryId = it.id!!
            val foundEntity = DataProvider().readByEntryId(edmEntitySet, mockEntryId)!!
            val entryId = "${foundEntity.getProperty(idPropertyName).value}"
            val queryId = "$idPropertyValue".trim('"')

            if (entryId == queryId) {
                DataContainer.removeEntry(mockEntryId)
                entryIds.add(mockEntryId)
            }
        }
```

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: src/main/kotlin/fi/metatavu/odata/mock/data/DataProvider.kt (hunk 1 above)
Comment (platonovsimeon) [reply src/main/kotlin/fi/metatavu/odata/mock/data/DataProvider.kt]: Fixed

**Pull request merged**